### PR TITLE
Make struct "ChunkSize" constructor explicit to avoid implicit construction in RangePolicy

### PR DESCRIPTION
**Pull request by tcclevenger**
Release 4.3 unintentionally introduced the ability for `RangePolicy` constructor to take as it's final argument any type convertible to `ChunkSize` and implicitly construct a `ChunkSize` object for the input. Remove that by making the `ChunkSize(int)` constructor explicit.

### CODE REVIEW

**Comment by tcclevenger** [comment]
@dalg24 I'm trying to find a clean way to deprecate the old behavior. Is this necessary? I'm guessing in a single release cycle not too high of a chance that it is being used unintentionally.

**Comment by dalg24** [comment]
When I said "investigate" in https://github.com/kokkos/kokkos/issues/7142#issuecomment-2233767695, I meant search downstream (Trilinos?) whether code breaks, since as Nevin pointed out, it is a breaking change even though we never intended to let someone construct `RangePolicy([exec,] (integral)begin, (integral)end, (integral)chunk_size)`.
Considering that we only accidentally enabled that in 4.3, usage should be limited.
A quick search in Trilinos yields
```
packages/stokhos/src/sacado/kokkos/Stokhos_Tpetra_Utilities.hpp:343:      const size_t ChunkSize = 16;
packages/stokhos/src/sacado/kokkos/Stokhos_Tpetra_Utilities.hpp:344:      const size_t M = (m+ChunkSize-1)/ChunkSize;
packages/stokhos/src/sacado/kokkos/Stokhos_Tpetra_Utilities.hpp:346:      Policy policy(M,ChunkSize,ChunkSize);
```
But interestingly, this code was committed in 2018 which means we need to find out if that code was actually built and if so why it worked(?)

**Comment by dalg24** [comment]
If we decide (I expect you will make recommendations) we need to go through deprecations, we could do something along these lines https://godbolt.org/z/6TWrfx78a

**Comment by tcclevenger** [comment]
In those Trilinos files, the `Policy` is a `TeamPolicy`, it is just a coincidence that "chunk size" is used as a variable name. I'll do a little more searching in Trilinos.

**Comment by dalg24** [comment]
> In those Trilinos files, the `Policy` is a `TeamPolicy`, it is just a coincidence that "chunk size" is used as a variable name. I'll do a little more searching in Trilinos.

That makes more sense.  Thanks.

**Comment by dalg24** [comment]
> If we decide (I expect you will make recommendations) we need to go through deprecations, we could do something along these lines https://godbolt.org/z/6TWrfx78a

I just realized this won't work. `ChunkSize cs{1};` will be ambiguous.

**Comment by tcclevenger** [comment]
Yeah, I'm testing now we need something different.

**Review by masterleinad — APPROVED** [review]
We'll need to update the documentation accordingly. 
I doubt that anyone noticed any difference but I think it's fair to make the constructor explicit either way.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 1]
```diff
@@ -40,7 +40,12 @@ struct ParallelReduceTag {};
 
 struct ChunkSize {
   int value;
+  explicit ChunkSize(int value_) : value(value_) {}
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+  template <typename T = void>
```
Comment (nliber) [inline core/src/Kokkos_ExecPolicy.hpp:45]: I didn't know that would work...

**Review by nliber — APPROVED** [review]
So, I tried the equivalent of this, but the RangePolicy constructor would now implicitly call the explicit ChunkSize constructor.  I don't know why (both under clang 18.1 and gcc 14.1).

**Review by dalg24 — CHANGES_REQUESTED** [review]
Add a compile time test that check that when deprecated code is disabled, one cannot build a RangePolicy with a trailing integral argument

**Review by nmm0 — APPROVED** [review]
Thanks for doing this, implicit conversions are pretty unfortunate here

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Please make sure you list it under "incompatibilities" in the change log

**Pull request merged**